Sync receiver

trevorgowing opened the pull request.

trevorgowing committed:
Add Database.replaceLocalWithRemote

* In order to support peer to peer synchronization local resource are
  changes are required to be overwritten. Conflict resolution will be
  handled upstream through an implementation of
  ConflictResolutionStrategy.

trevorgowing commented:
@bausmeier Were you thinking that the `SyncSender`/`SyncReceiver` would use an/the instance of `FhirEngine` or were you thinking of using an/the instance of Database rather?

Your comment about the transactions made me think about at what level these services should be operating.

The reason I ask comes back to my original question, in trying to determine at what level to make the decision about whether to perform the lookup of the resource at the service or Database level.

Should our services be working with Database directly, I am thinking that maybe it makes sense to expose the transaction management from that API so that the functions can continue to perform a single operation but we can wrap those operations in a transaction from the service.

But if we are dealing with an/the instance of `FhirEngine` then I don't think that makes sense since it would mean leaking implementation concepts out of the API.

bausmeier commented:
Looking at what currently lives in the `FhirEngine` vs in the `Database`, and how the server sync works, I think we'll need to be working at a lower level than at the engine - so at the database. I agree that the transaction handling could be exposed from the database as long as it's still encapsulated within the engine.

trevorgowing commented:
@bausmeier We need to add a dependency to the `p2p` module on `org.jetbrains.kotlinx:kotlinx-coroutines-core` which we are both going to need. Should I add that to the `sync-payload` branch?

bausmeier commented:
Yeah you can. And I think we should go ahead and merge that PR if you're happy with it?

trevorgowing commented:
@bausmeier I am guessing that we will need the corresponding test library too? Although I see that `engine` has a dependency on the test library so I guess we will inherit it from there. I haven't gotten to writing tests yet.

bausmeier commented:
I'm not familiar with the testing library so I'm not sure. Sounds likely that we'd need to add it. The tests I wrote don't use coroutines.

trevorgowing committed:
Add Database.withTransaction

* Support the wrapping of multiple operations at the service level in a
  transaction.
* In the context of peer to peer sync, resources need to be looked up
  and it determined whether they should be inserted into or updated in the
  database. In order to continue to follow the pattern of the existing
  Database functions which perform a single operation, transaction
  management needs to be exposed to the service level.

trevorgowing committed:
Add ConflictResolutionStrategy interface to p2p

* Enable conflicts to be resolved using the specified
  ConflictResolutionStategy implementation.

trevorgowing commented:
> Looking at what currently lives in the `FhirEngine` vs in the `Database`, and how the server sync works, I think we'll need to be working at a lower level than at the engine - so at the database. I agree that the transaction handling could be exposed from the database as long as it's still encapsulated within the engine.

I only just noticed that [Database](https://github.com/decode-development/android-fhir/blob/p2p-sync/engine/src/main/java/com/google/android/fhir/db/Database.kt) is `internal`. So it seems to me like the intention of the engine was for all interaction to be through [FhirEngine](https://github.com/decode-development/android-fhir/blob/p2p-sync/engine/src/main/java/com/google/android/fhir/FhirEngine.kt).

Perhaps, we should reconsider our discussion about working with `Database` directly. Or at least I think that we should add it as an agenda item for discussion in the call tomorrow.

What do you think?

We could implement what would have been implemented in `SyncReceiver` in `FhirEngine`.

I do prefer your design though. I just think we need to clear publicizing `Database` with the rest of the team since it would break what seems like a fundamental design decision.

trevorgowing committed:
Publicize Database

* In order to support peer to peer synchronization a lower level of
  control is required than was originally intended for FhirEngine.

bausmeier commented:
Ah, yeah that's a good point. When I was experimenting with the API originally I was doing so from within the `engine` module, but I'd like to keep the p2p stuff separate from that. The `Database` class should probably be kept internal then, and the functions we need would have to be accessed via `FhirEngine`. I don't think that publicising `Database` is the right approach. Raising this on the call tomorrow to get some suggestions is a good idea.

trevorgowing commented:
@bausmeier We could make `SyncReceiver` an interface and introduce a new `FhirEngine` interface which combines the existing one with `SyncReceiver` and supply a corresponding implementation.

Therefore we keep the existing `FhirEngine` unchanged and it is up to the client to instantiate the `FhirEngine` implementation with the appropriate support for the functionality they require.
